### PR TITLE
PLT-6907 Fixed execution cost analysis of merkleized contracts

### PR DESCRIPTION
**Pull request by bwbush**
Fixed deep demerkleization of contract for the Plutus execution-cost analysis, and added golden tests for that.

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [x] [Test report is updated](https://github.com/input-output-hk/marlowe-cardano/blob/main/marlowe/test/test-report.md) (if relevant)
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
        - Review required
        - [x] Substantial changes to code, test, or documentation
        - [ ] Change made to Marlowe validator (@bwbush and @palas must be included as reviewers)
        - Review not required
        - [ ] Minor changes to non-critical code, documentation, nix derivations, configuration files, or scripts
        - [ ] Formatting, spelling, grammar, or reorganization
    - [x] Reviewer requested

### CODE REVIEW

**Review by paluh — COMMENTED** [review]
May I ask - so the critical fix can be distilled to:

```diff
- demerkleizeCase c = pure c
+ demerkleizeCase (Case action continuation) = Case action <$> demerkleize requireContinuations f continuation
```
?

**Comment by bwbush** [comment]
> May I ask - so the critical fix can be distilled to:
> 
> ```diff
> - demerkleizeCase c = pure c
> + demerkleizeCase (Case action continuation) = Case action <$> demerkleize requireContinuations f continuation
> ```

Yes, pretty much.

File: marlowe/src/Language/Marlowe/Core/V1/Merkle.hs [hunk 1]
```diff
@@ -127,45 +126,66 @@ merkleize' f (When cases timeout contract) = When <$> mapM merkleizeCase cases <
...
-        pure $ Case action continuation
-    demerkleizeCase c = pure c
+        continuation' <- asks . M.lookup $ DatumHash mh
+        case continuation' of
+          Just continuation ->
+            fmap (Case action) . demerkleize requireContinuations f =<< f continuation
+          Nothing ->
+            if requireContinuations
+              then throwError . fromString $ "Missing continuation for hash " <> show mh <> "."
+              else pure mc
```
Comment (paluh) [inline marlowe/src/Language/Marlowe/Core/V1/Merkle.hs:180]: Cosmetics: some remaining commented code.

File: marlowe/src/Language/Marlowe/Core/V1/Merkle.hs [hunk 2]
```diff
@@ -127,45 +126,66 @@ merkleize' f (When cases timeout contract) = When <$> mapM merkleizeCase cases <
...
 
 -- | Demerkleize selected case statements in a contract.
-demerkleize'
-  :: (IsString e)
+demerkleize
+  :: forall e m
+   . (IsString e)
   => (MonadError e m)
-  => (Contract -> ReaderT Continuations m Contract)
+  => Bool
```
Comment (paluh) [inline marlowe/src/Language/Marlowe/Core/V1/Merkle.hs:154]: Cosmetics: That boolean alone is not really problematic (but harder to read on the call site). Let's say that when we have second `Bool` signature then we will introduce `newtype`s ;-)

**Review by paluh — APPROVED** [review]
Looks great. I like that test case - I will probably port it to our DSL based test suite ;-)

**Pull request merged**